keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,833 @@
+	// Webhook implementations are encouraged to ignore rawSelector.
+    // The kube-apiserver's SubjectAccessReview will parse the rawSelector. 
+	RawSelector string
+
+	// requirements is the parsed interpretation of a field selector.
+	// All requirements must be met for a resource instance to match the selector.
+	// Webhook implementations should handle requirements, but how to handle them is up to the webhook.
+	// Since requirements can only limit the request, it is safe to authorize as unlimited request if the requirements
+	// are not understood.
+	Requirements []FieldSelectorRequirement
+}
+
+// LabelSelectorAttributes indicates a field limited access.
+// For webhooks:
+// The kube-apiserver will never send a request with rawSelector set, but we cannot control what other clients directly send.
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the request is not limited.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
+// For the kube-apiserver:
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
+type LabelSelectorAttributes struct {
+	// rawSelector is the serialization of a field selector that would be included in a query parameter.
+    // Webhook implementations are encouraged to ignore rawSelector.
+	// The kube-apiserver's SubjectAccessReview will parse the rawSelector. 
+	RawSelector string
+
+    // requirements is the parsed interpretation of a label selector.
+    // All requirements must be met for a resource instance to match the selector.
+    // Webhook implementations should handle requirements, but how to handle them is up to the webhook.
+    // Since requirements can only limit the request, it is safe to authorize as unlimited request if the requirements
+    // are not understood.
+	Requirements []metav1.LabelSelectorRequirement
+}
+
+type FieldSelectorRequirement struct {
+	// key is the label key that the selector applies to.
+	Key string `json:"key" protobuf:"bytes,1,opt,name=key"`
+	// operator represents a key's relationship to a set of values.
+	// Valid operators are In, NotIn, Exists, DoesNotExist
+	// The list of operators may grow in the future.
+	// Webhook authors are encouraged to ignore unrecognized operators and assume they don't limit the request.
+	// The semantics of "all requirements are AND'd will not change, so other requirements can continue to be enforced.
```

Comment:
I believe we should provide a function to "simplify" the requirements into a "canonical representation". For example, if there are multiple rules for the same key, one can apply the following rules:

1. Simplify Equals/NotEquals to In/NotIn (this already happens in the rawSelector parsing layer essentially)
Equals `foo` => In `[foo]`
NotEquals `foo` => NotIn `[foo]`

2. Simplify In/NotIn separately

In `[a,b]` && In `[b,c]` => In `[b]` (intersection)
NotIn `[a,b]` && NotIn `[b,c]` => NotIn `[a,b,c]` (union)

By this stage, we have at most one In and one NotIn requirement per key (plus any possible Exists/NotExist, handled later)

3. Simplify In/NotIn together

In `[a,b]` && NotIn `[b,c]` => In `[a]` (subtraction)

Note: this must be done last (as far as I can see), as the previous union/intersection operations are commutative, but subtraction is not.

In the end of this process we have at most one requirement with operator In or NotIn (not both), plus any possible Exists/NotExist, handled later.

4. Simplify Exists/NotExists

Exists && Exists => Exists
NotExists && NotExists => NotExists
Exists && NotExists => ∅ (a label cannot both be defined and not for the same object, unsatisfiable)

At this point, we have at most one (In or NotIn) and one (Exists or NotExists).
Of course, we can short-circuit earlier if we get an empty set before, as then the requirements are unsatisfiable by any object.

5. Combine one (In or NotIn) and one (Exists or NotExists)

In `[a,b]` && Exists => In `[a,b]` (if foo=a or foo=b is true, foo does exist)
NotIn `[a,b]` && Exists => NotIn `[a,b]` && Exists (no simplification possible)
In `[a,b]` && NotExists => ∅ (if the label must both be set and not set, req is unsatisfiable)
NotIn `[a,b]` && NotExists => NotExists (a label not existing certainly is not set to a given value)

At this point, of all possible requirements, we have at most the following requirements for each key:
In `[a,b]`
NotIn `[a,b]`
Exists
NotExists
∅ (unsatisfiable, which makes the whole requirements list for any object unsatisfiable)
NotIn `[a,b]` && Exists

This "simplified/parsed" requirements set could provide a dedicated interface, which can answer e.g. `ObjectMatches(obj)`, `Unsatisfiable()`, `Equal(otherRequirements)`, `GreaterThan(otherRequirements`, etc., such that one can semantically compare varying requirements sets with each other (a use case that will be very much needed).

Reply:
> I believe we should provide a function to "simplify" the requirements into a "canonical representation". For example, if there are multiple rules for the same key, one can apply the following rules:

I don't see a reason to do this at the moment.

Reply:
Yeah. Not needed for this KEP indeed, but I have some use-cases where this would be useful, so I can experiment with that outside of core indeed.